pages/blog/posts/understanding-lexical-dynamic-scopes.md

```diff
@@ -0,0 +1,418 @@
+applicator into the subschema at `/properties/name`. This subschema introduces
+a new schema resource. Therefore, the dynamic scope now consists of *both* the
+root schema resource and the nested schema resource called
+`https://example.com/name`, in order:
+
+![The dynamic scope of a JSON Schema (2)](/img/posts/2024/understanding-lexical-dynamic-scopes/dynamic-scope-2.webp)
+
+In comparison to the lexical scope, the dynamic scope of a schema cannot always
+be statically determined, as the evaluation path often depends on the instance.
+For example, for schemas that make use of logic applicator keywords such as
+`if` or `oneOf`, the ordered sequence of schema resources in scope may vary
+depending on the characteristics of the instance.
+
+### Following References
+
+So far, we learnt that for the lexical scope, following a reference consists in
+abandoning the lexical scope of the origin schema and entering the lexical
+scope of the destination schema. In comparison, for the dynamic scope,
+following a reference to another schema resource involves *retaining* the
+current dynamic scope and *pushing* the destination schema resource to the top
+of the stack.
+
+#### Within Schema Resources
+
+Just like with the lexical scope, if a reference points to a subschema within
+the same schema resource, the dynamic scope remains the same. In other words,
+if the destination schema resource is the same as the schema resource at the
+top of the stack, the dynamic scope is not modified. Therefore, until the
+evaluation process encounters a reference to another schema resource (either
+local or remote), *the lexical scope and the dynamic scope align*:
+
+![Dynamic scope and lexical scopes sometimes align](/img/posts/2024/understanding-lexical-dynamic-scopes/dynamic-scope-internal-align.webp)
+
+#### Across Schema Resources
+
+Leaving the simple case behind, lets consider an example consisting of local
+and remote references across schema resources. In the top left, an example
+instance and a root schema resource named `https://example.com` that declares
+two nested schema resources: `https://example.com/name` (at `/properties/name`)
+and `https://example.com/person` (at `/$defs/person`) where the former
+references the latter (from `/properties/name/$ref`). Furthermore,
+`https://example.com/person` references a nested schema resource called
+`https://example.com/item` (from `/$defs/person/$ref`) that is part of an
+external schema resource called `https://example.com/people` shown in the
+bottom left. In the right, a directed graph representation of the relationship
+between these schema resources and the dynamic scope.
+
+Like the other examples so far, the evaluation process starts with the top
+level schema. The dynamic scope at that point is the root schema resource, and
+all other schema resources are out of scope:
+
+![The dynamic scope and remote references (1)](/img/posts/2024/understanding-lexical-dynamic-scopes/dynamic-scope-ref-remote-1.webp)
```

Comment:
> `"$ref": "people/#items"

We usually advise against referencing directly into other keywords like this.  The advice is generally to extract the target into `$defs` or to an external schema and reference from both locations.

I recognize that you have limited space here, though.

Reply:
Also since that subschema defines an `$id`, this is using a pointer reference across resource boundaries, which is undefined.  (Most implementations support it, but it's not a requirement to support.)

The target subschema here has a resolved ID of `https://example.com/item` and the reference can just be changed to `"$ref": "item"`, which feels weird.

Reply:
Would it make it nicer if I set an anchor at `item` and then reference i.e. `people#item`?

Reply:
If you do that, I think you lose your example where you're skipping over `people`.

Will continue to follow up in the [other thread](https://github.com/json-schema-org/website/pull/274#discussion_r1482235743).

Reply:
I think its OK, as even with `people#item`, `people` is never at the dynamic scope, right?

Reply:
Right. Let's continue the discussion on the other thread :+1: